File: core/src/main/java/com/tvsc/core/exception/ExceptionUtil.java
package com.tvsc.core.exception;


import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author Taras Zubrei
 */
public class ExceptionUtil {

    private static Logger LOGGER = LoggerFactory.getLogger(ExceptionUtil.class);

    public static <T> T wrapCheckedException(StatementWithReturnValue<T> statement, ApplicationException exception) {
        try {
            return statement.evaluate();
        } catch (RuntimeException e) {
            exception.afterThrowing();
            throw e;
        } catch (Throwable e) {
            exception.setCause(e);
            exception.afterThrowing();
            throw exception;
        }
    }
}

File: core/src/main/kotlin/com/tvsc/core/exception/ExceptionUtil.kt
package com.tvsc.core.exception


import java.util.concurrent.Callable

/**
 * @author Taras Zubrei
 */
object ExceptionUtil {

    @JvmStatic
    fun <T, K> wrapCheckedException(exception: K, statement: Callable<T>): T where K : ApplicationException {
        try {
            return statement.call()
        } catch (e: RuntimeException) {
            throw e
        } catch (e: Throwable) {
            val constructor = exception.javaClass.getConstructor(String::class.java, Throwable::class.java)
            throw constructor.newInstance(exception.message, e)
        }
    }
}
